src/stan/model/indexing/assign_cl.hpp

```diff
@@ -0,0 +1,385 @@
+      [expr_lhs, expr_rhs, row_index, name, prev_vals]() mutable {
+        auto&& lhs_val = rvalue(expr_lhs.val_op(), name, row_index);
+        auto&& lhs_adj_holder = rvalue(expr_lhs.adj(), name, row_index);
+        auto& lhs_adj = lhs_adj_holder;
+
+        math::results(lhs_val, expr_rhs.adj(), lhs_adj) = math::expressions(
+            prev_vals, expr_rhs.adj() + lhs_adj,
+            math::constant(0.0, lhs_adj.rows(), lhs_adj.cols()));
+      });
+}
+// the "forwarding" overload
+inline void assign(math::var_value<math::matrix_cl<double>>& expr_lhs,
+                   math::var_value<math::matrix_cl<double>>&& expr_rhs,
+                   const char* /*name*/, index_omni /*row_index*/) {
+  expr_lhs.vi_ = expr_rhs.vi_;
+}
+
+/**
+ * Assign one reverse mode kernel generator expression to another, using given
+ * indices.
+ *
+ * @tparam ExprLhs type of the expression on the left hand side of the
+ * assignment
+ * @tparam ExprRhs type of the expression on the right hand side of the
+ * assignment
+ * @tparam RowIndex type of row index
+ * @tparam ColIndex type of columnindex
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param expr_rhs expression on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing rows of `expr_lhs`
+ * @param col_index index used for indexing columns of `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ * @throw std::invalid_argument If the right hand side size isn't the same as
+ * the indexed left hand side size.
+ */
+template <typename ExprLhs, typename ExprRhs, typename RowIndex,
+          typename ColIndex,
+          require_all_rev_kernel_expression_t<ExprLhs, ExprRhs>* = nullptr,
+          require_any_not_same_t<RowIndex, ColIndex, index_uni>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ExprRhs& expr_rhs,
+                   const char* name, const RowIndex& row_index,
+                   const ColIndex& col_index) {
+  if (std::is_same<RowIndex, index_omni>::value
+      && std::is_same<ColIndex, index_omni>::value) {
+    stan::math::check_size_match("omni assign", "left hand side rows",
+                                 expr_lhs.rows(), name, expr_rhs.rows());
+    stan::math::check_size_match("omni assign", "left hand side columns",
+                                 expr_lhs.cols(), name, expr_rhs.cols());
+  }
+  decltype(auto) lhs = rvalue(expr_lhs.val_op(), name, row_index, col_index);
+  math::arena_matrix_cl<double> prev_vals = lhs;
+  lhs = expr_rhs.val();  // assign the values
+  math::reverse_pass_callback([expr_lhs, expr_rhs, row_index, col_index, name,
+                               prev_vals]() mutable {
+    auto&& lhs_val_holder
+        = rvalue(expr_lhs.val_op(), name, row_index, col_index);
+    auto&& lhs_adj_holder = rvalue(expr_lhs.adj(), name, row_index, col_index);
+    auto& lhs_val = lhs_val_holder;
+    auto& lhs_adj = lhs_adj_holder;
```

Comment:
Why does there need to be `lhs_val_holder` and `lhs_val`? I usually don't like comments near code, but If there's some precise mechanics going on here it would be nice to have something written down either here or in the doc comment

Reply:
I forgot to replace `auto&&` with `decltype(auto)` in some places. Fixed.